File: deadchest-core/src/main/java/me/crylonz/deadchest/commands/DCCommandRegistrationService.java
package me.crylonz.deadchest.commands;

import me.crylonz.deadchest.ChestData;
import me.crylonz.deadchest.DeadChestLoader;
import me.crylonz.deadchest.Permission;
import me.crylonz.deadchest.db.InMemoryChestStore;
import me.crylonz.deadchest.utils.ConfigKey;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.entity.ArmorStand;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import java.util.Collection;
import java.util.Date;
import java.util.Map;

import static me.crylonz.deadchest.DeadChestLoader.*;
import static me.crylonz.deadchest.DeadChestManager.cleanAllDeadChests;
import static me.crylonz.deadchest.DeadChestManager.removeDeadChest;
import static me.crylonz.deadchest.db.IgnoreItemListRepository.loadIgnoreIntoInventory;

public class DCCommandRegistrationService extends DCCommandRegistration {

    public DCCommandRegistrationService(DeadChestLoader plugin) {
        super(plugin);
    }

    public void registerReload() {
        registerCommand("dc reload", Permission.ADMIN.label, () -> {
            loadIgnoreIntoInventory(ignoreList);
            DeadChestLoader.plugin.reloadConfig();
            plugin.registerConfig();
            local.reloadLanguage(config.getString(ConfigKey.LOCALIZATION_LANGUAGE));

            sender.sendMessage(local.prefixed("commands.reload.success"));
        });
    }

    public void registerRepairForce() {
        registerCommand("dc repair force", Permission.ADMIN.label, () -> repair(true));
    }

    public void registerRepair() {
        registerCommand("dc repair", Permission.ADMIN.label, () -> repair(false));
    }

    private void repair(Boolean forced) {
        if (player == null) {
            sender.sendMessage(local.prefixed("commands.error.player-only"));
            return;
        }

        getSchedulerAdapter().executeForEntity(player, () -> {
            Collection<Entity> entities = player.getWorld().getNearbyEntities(player.getLocation(), 100.0D, 25.0D, 100.0D);
            int holoRemoved = 0;

            for (Entity entity : entities) {
                if (entity.getType() != EntityType.ARMOR_STAND) {
                    continue;
                }

                ArmorStand armorStand = (ArmorStand) entity;
                if (armorStand.hasMetadata("deadchest") || forced) {
                    holoRemoved++;
                    entity.remove();
                }
            }

            player.sendMessage(local.prefixed("commands.operation.holograms-removed", holoRemoved));
        });
    }

    public void registerRemoveInfinite() {
        registerCommand("dc removeinfinite", Permission.ADMIN.label, () -> {
            int count = 0;
            final InMemoryChestStore inMemoryChestStore = DeadChestLoader.getChestDataCache();
            final Map<Location, ChestData> chestDataMap = inMemoryChestStore.getAllChestData();

            if (chestDataMap != null && !chestDataMap.isEmpty()) {
                for (final ChestData chestData : chestDataMap.values()) {
                    if (chestData.getChestLocation().getWorld() != null &&
                            (chestData.isInfinity() || config.getInt(ConfigKey.DEADCHEST_DURATION) == 0)) {
                        getSchedulerAdapter().executeAtLocation(chestData.getChestLocation(), () -> removeDeadChest(chestData));
                        count++;
                    }
                }
            }

            sender.sendMessage(local.prefixed("commands.operation.deadchests-removed", count));
        });
    }

    public void registerRemoveAll() {
        registerCommand("dc removeall", Permission.ADMIN.label, () -> {
            int count = cleanAllDeadChests();
            sender.sendMessage(local.prefixed("commands.operation.deadchests-removed", count));
        });
    }

    public void registerRemoveOwn() {
        registerCommand("dc remove", Permission.REMOVE_OWN.label, () -> {
            if (player != null) {
                removeAllDeadChestOfPlayer(player.getName());
            } else {
                sender.sendMessage(local.prefixed("commands.error.player-only"));
            }
        });
    }

    public void registerRemoveOther() {
        registerCommand("dc remove {0}", Permission.REMOVE_OTHER.label, () -> removeAllDeadChestOfPlayer(args[1]));
    }

    private void removeAllDeadChestOfPlayer(String playerName) {
        int count = 0;
        final InMemoryChestStore chestDataCache = getChestDataCache();
        final Map<Location, ChestData> chestDataList = chestDataCache.getAllChestData();

        if (chestDataList != null && !chestDataList.isEmpty()) {
            for (ChestData chestData : chestDataList.values()) {
                if (chestData != null && chestData.getPlayerName().equalsIgnoreCase(playerName)) {
                    getSchedulerAdapter().executeAtLocation(chestData.getChestLocation(), () -> removeDeadChest(chestData));
                    count++;
                }
            }
        }

        sender.sendMessage(local.prefixed("commands.operation.deadchests-removed-player", count, playerName));
    }

    public void registerListOwn() {
        registerCommand("dc list", null, () -> {
            if (player == null) {
                sender.sendMessage(local.prefixed("commands.error.player-only"));
                return;
            }

            if (player.hasPermission(Permission.LIST_OWN.label) || !config.getBoolean(ConfigKey.REQUIRE_PERMISSION_TO_LIST_OWN)) {
                Date now = new Date();
                final Map<Location, ChestData> chestDataList = getChestDataCache().getAllChestData();
                if (!chestDataList.isEmpty()) {
                    sender.sendMessage(local.prefixed("commands.list.title.own"));
                    for (ChestData data : chestDataList.values()) {
                        if (data.getPlayerUUID().equals(player.getUniqueId())) {
                            displayChestData(now, data);
                        }
                    }
                } else {
                    player.sendMessage(local.prefixed("commands.list.none.player"));
                }
            }
        });
    }

    public void registerListOther() {
        registerCommand("dc list {0}", Permission.LIST_OTHER.label, () -> {
            Date now = new Date();
            final Map<Location, ChestData> chestDataList = getChestDataCache().getAllChestData();

            if (args[1].equalsIgnoreCase("all")) {
                if (!chestDataList.isEmpty()) {
                    sender.sendMessage(local.prefixed("commands.list.title.all"));
                    for (ChestData data : chestDataList.values()) {
                        displayChestData(now, data);
                    }
                } else {
                    sender.sendMessage(local.prefixed("commands.list.none.global"));
                }
                return;
            }

            if (!chestDataList.isEmpty()) {
                sender.sendMessage(local.prefixed("commands.list.title.player", args[1]));
                for (ChestData data : chestDataList.values()) {
                    if (data.getPlayerName().equalsIgnoreCase(args[1])) {
                        displayChestData(now, data);
                    }
                }
            } else {
                sender.sendMessage(local.prefixed("commands.list.none.global"));
            }
        });
    }

    private void displayChestData(Date now, ChestData chestData) {
        String worldName = chestData.getChestLocation().getWorld() != null ?
                chestData.getChestLocation().getWorld().getName() : local.get("common.unknown-world");

        if (chestData.isInfinity() || config.getInt(ConfigKey.DEADCHEST_DURATION) == 0) {
            sender.sendMessage(local.format(
                    "commands.list.entry.infinity",
                    worldName,
                    chestData.getChestLocation().getX(),
                    chestData.getChestLocation().getY(),
                    chestData.getChestLocation().getZ(),
                    local.get("chest.time-left")
            ));
            return;
        }

        long diff = now.getTime() - (chestData.getChestDate().getTime() + config.getInt(ConfigKey.DEADCHEST_DURATION) * 1000L);
        long diffSeconds = Math.abs(diff / 1000 % 60);
        long diffMinutes = Math.abs(diff / (60 * 1000) % 60);
        long diffHours = Math.abs(diff / (60 * 60 * 1000));

        sender.sendMessage(local.format(
                "commands.list.entry.timed",
                chestData.getChestLocation().getX(),
                chestData.getChestLocation().getY(),
                chestData.getChestLocation().getZ(),
                diffHours,
                diffMinutes,
                diffSeconds,
                local.get("chest.time-left")
        ));
    }

    public void registerGiveBack() {
        registerCommand("dc giveback {0}", Permission.GIVEBACK.label, () -> {
            Player targetPlayer = null;
            final InMemoryChestStore inMemoryChestStore = DeadChestLoader.getChestDataCache();
            final Map<Location, ChestData> chestDataList = inMemoryChestStore.getAllChestData();

            for (ChestData data : chestDataList.values()) {
                if (!data.getPlayerName().equalsIgnoreCase(args[1])) {
                    continue;
                }

                targetPlayer = Bukkit.getPlayer(data.getPlayerUUID());
                if (targetPlayer != null && player != null && player.isOnline()) {
                    final Player finalTargetPlayer = targetPlayer;
                    getSchedulerAdapter().executeForEntity(finalTargetPlayer, () -> {
                        for (ItemStack itemStack : data.getInventory()) {
                            if (itemStack != null) {
                                finalTargetPlayer.getWorld().dropItemNaturally(finalTargetPlayer.getLocation(), itemStack);
                            }
                        }
                    });
                    getSchedulerAdapter().executeAtLocation(data.getChestLocation(), () -> removeDeadChest(data));
                }
                break;
            }

            if (targetPlayer != null) {
                sender.sendMessage(local.prefixed("commands.giveback.success.sender"));
                targetPlayer.sendMessage(local.prefixed("commands.giveback.success.target"));
            } else {
                sender.sendMessage(local.prefixed("commands.giveback.target-not-found"));
            }
        });
    }

    public void registerIgnoreList() {
        registerCommand("dc ignore ", Permission.ADMIN.label, () ->
                getSchedulerAdapter().executeForEntity(player, () -> player.openInventory(ignoreList))
        );
    }
}

File: deadchest-core/src/main/kotlin/me/crylonz/deadchest/commands/DCCommandRegistrationService.kt
package me.crylonz.deadchest.commands

import me.crylonz.deadchest.ChestData
import me.crylonz.deadchest.DeadChest
import me.crylonz.deadchest.DeadChest.Companion.chestData
import me.crylonz.deadchest.DeadChest.Companion.dcConfig
import me.crylonz.deadchest.DeadChest.Companion.fileManager
import me.crylonz.deadchest.DeadChest.Companion.local
import me.crylonz.deadchest.DeadChestManager
import me.crylonz.deadchest.Permission
import me.crylonz.deadchest.Permission.*
import me.crylonz.deadchest.utils.ConfigKey.DEADCHEST_DURATION
import me.crylonz.deadchest.utils.ConfigKey.REQUIRE_PERMISSION_TO_LIST_OWN
import org.bukkit.Bukkit
import org.bukkit.ChatColor.*
import org.bukkit.Material
import org.bukkit.entity.ArmorStand
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import java.util.*
import kotlin.math.abs


class DCCommandRegistrationService(plugin: DeadChest) : DCCommandRegistration(plugin) {
    fun registerReload() {
        registerCommand("dc reload", ADMIN.label) {
            fileManager.reloadChestDataConfig()
            fileManager.reloadLocalizationConfig()

            plugin.reloadConfig()
            plugin.registerConfig()

            fileManager.chestDataConfig.getList("chestData")
                ?.let { chestData = it as MutableList<ChestData> }

            local.set(
                fileManager.localizationConfig.getConfigurationSection("localisation")!!
                    .getValues(true)
            )
            sender.sendMessage(GREEN.toString() + local.get("loc_prefix") + " Plugin reload successfully")
        }
    }

    fun registerRepairForce() {
        registerCommand("dc repair force", ADMIN.label) { repair(true) }
    }

    fun registerRepair() {
        registerCommand("dc repair", ADMIN.label) { repair(false) }
    }

    private fun repair(forced: Boolean) {

        player?.let { player ->
            var holoRemoved = 0
            player.world.getNearbyEntities(player.location, 100.0, 25.0, 100.0)
                .filter { it.type == EntityType.ARMOR_STAND }
                .map { it as ArmorStand }
                .filter { it.hasMetadata("deadchest") || forced }
                .forEach {
                    holoRemoved++
                    it.remove()
                }

            sender.sendMessage("${local.get("loc_prefix")}$GOLD Operation complete. [$holoRemoved] hologram(s) removed")

        } ?: sender.sendMessage("${local.get("loc_prefix")}$RED Command must be called by a player")
    }

    fun registerRemoveInfinite() {
        registerCommand("dc removeinfinite", ADMIN.label) {

            var cpt = 0
            if (chestData.isNotEmpty()) {

                chestData
                    .filter { it.chestLocation.world != null }
                    .forEach {
                        if (it.isInfinity || dcConfig.getInt(DEADCHEST_DURATION) == 0) {
                            it.chestLocation.world!!.getBlockAt(it.chestLocation).type = Material.AIR
                            it.removeArmorStand()
                            chestData.remove(it)
                            cpt++
                        }
                    }
                fileManager.saveModification()
            }

            sender.sendMessage(
                "${local.get("loc_prefix")}${GOLD}Operation complete. [$GREEN$cpt$GOLD] deadchest(s) removed"
            )
        }
    }

    fun registerRemoveAll() {
        registerCommand("dc removeall", ADMIN.label) {
            val cpt: Int = DeadChestManager.cleanAllDeadChests()
            sender.sendMessage(
                ("${local.get("loc_prefix")}${GOLD}Operation complete. [$GREEN$cpt$GOLD] deadchest(s) removed")
            )
        }
    }

    fun registerRemoveOwn() {
        registerCommand("dc remove", Permission.REMOVE_OWN.label) {
            player?.let {
                removeAllDeadChestOfPlayer(it.name)
            } ?: sender.sendMessage("${local.get("loc_prefix")}${RED}Command must be called by a player")
        }
    }

    fun registerRemoveOther() {
        registerCommand("dc remove {0}", Permission.REMOVE_OTHER.label) {
            removeAllDeadChestOfPlayer(args[1])
        }
    }

    private fun removeAllDeadChestOfPlayer(playerName: String) {
        var cpt = 0
        if (chestData.isNotEmpty()) {

            chestData
                .filter { it.chestLocation.world != null }
                .filter { it.playerName.equals(playerName, ignoreCase = true) }
                .forEach {
                    it.chestLocation.world!!.getBlockAt(it.chestLocation).type = Material.AIR
                    it.removeArmorStand()
                    chestData.remove(it)
                    cpt++
                }
            fileManager.saveModification()
        }
        sender.sendMessage(
            local.get("loc_prefix") + GOLD + "Operation complete. [" + GREEN + cpt + GOLD + "] deadchest(s) removed of player " + playerName
        )
    }

    fun registerListOwn() {
        registerCommand("dc list", null) {
            player?.let {
                if (it.hasPermission(LIST_OWN.label) || !dcConfig.getBoolean(REQUIRE_PERMISSION_TO_LIST_OWN)) {
                    if (chestData.isNotEmpty()) {
                        sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_dclistown")} :")
                        chestData
                            .filter { data -> data.playerUUID.equals(it.uniqueId.toString(), ignoreCase = true) }
                            .forEach { data -> displayChestData(Date(), data) }
                    } else {
                        it.sendMessage("${local.get("loc_prefix")}${local.get("loc_nodc")}")
                    }
                }
            } ?: sender.sendMessage("${local.get("loc_prefix")}${RED}Command must be called by a player")
        }
    }

    fun registerListOther() {
        registerCommand("dc list {0}", LIST_OTHER.label) {

            val now = Date()
            if (args[1].equals("all", ignoreCase = true)) {
                if (chestData.isNotEmpty()) {
                    sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_dclistall")}:")
                    chestData.forEach { data: ChestData ->
                        displayChestData(now, data)
                    }
                } else {
                    sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_nodcs")}")
                }
            } else {
                if (chestData.isNotEmpty()) {
                    sender.sendMessage("${local.get("loc_prefix")}$GREEN${args[1]} deadchests :")
                    chestData
                        .filter { it.playerName.equals(args[1], ignoreCase = true) }
                        .forEach { displayChestData(now, it) }
                } else {
                    sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_nodcs")}")
                }
            }
        }
    }

    private fun displayChestData(now: Date, chestData: ChestData) {
        val worldName = when {
            chestData.chestLocation.world != null -> chestData.chestLocation.world!!.name
            else -> "???"
        }

        if (chestData.isInfinity || dcConfig.getInt(DEADCHEST_DURATION) == 0) {
            sender.sendMessage(
                "-" + AQUA + " World: " + WHITE + worldName + " |"
                        + AQUA + " X: " + WHITE + chestData.chestLocation.x
                        + AQUA + " Y: " + WHITE + chestData.chestLocation.y
                        + AQUA + " Z: " + WHITE + chestData.chestLocation.z
                        + " | "
                        + "∞ " + local.get("loc_endtimer")
            )
        } else {
            val diff = now.time - (chestData.chestDate.time + dcConfig.getInt(DEADCHEST_DURATION) * 1000L)
            val diffSeconds = abs(diff / 1000 % 60)
            val diffMinutes = abs(diff / (60 * 1000) % 60)
            val diffHours = abs(diff / (60 * 60 * 1000))
            sender.sendMessage(
                "-" + AQUA + " X: " + WHITE + chestData.chestLocation.x
                        + AQUA + " Y: " + WHITE + chestData.chestLocation.y
                        + AQUA + " Z: " + WHITE + chestData.chestLocation.z
                        + " | " +
                        +diffHours + "h "
                        + diffMinutes + "m "
                        + diffSeconds + "s " + local.get("loc_endtimer")
            )
        }
    }

    fun registerGiveBack() {
        registerCommand("dc giveback {0}", GIVEBACK.label) {

            var targetPlayer: Player? = null

            chestData
                .filter { it.playerName.equals(args[1], ignoreCase = true) }
                .forEach {
                    targetPlayer = Bukkit.getPlayer(UUID.fromString(it.playerUUID))
                    targetPlayer?.let { currentPlayer ->
                        if (currentPlayer.isOnline) {
                            it.inventory
                                .filterNotNull()
                                .forEach { itemStack ->
                                    currentPlayer.world.dropItemNaturally(currentPlayer.location, itemStack)
                                }
                            currentPlayer.world.getBlockAt(it.chestLocation).type = Material.AIR
                            it.removeArmorStand()
                            chestData.remove(it)
                        }
                    }

                }

            targetPlayer?.let {
                sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_dcgbsuccess")}")
                it.sendMessage("${local.get("loc_prefix")}${local.get("loc_gbplayer")}")
            } ?: sender.sendMessage("${local.get("loc_prefix")}${local.get("loc_givebackInfo")}")
        }
    }
}

